drivers/dma/dma_intel_adsp_hda.c

```diff
@@ -230,6 +230,10 @@ int intel_adsp_hda_dma_start(const struct device *dev, uint32_t channel)
 
 	__ASSERT(channel < cfg->dma_channels, "Channel does not exist");
 
+	if (intel_adsp_hda_is_enabled(cfg->base, cfg->regblock_size, channel)) {
```

Comment:
I guess this could be a common implement, can we put it to dam.h z_impl_dma_start and z_impl_dma_stop?  otherwise all other venders driver need update so

Reply:
Are you looking to have the driver API implement a function that start/stop can inspect then? That probably makes good sense

Reply:
right, and this API can have a default one, as we call start/stop in dma.h, and we can have internal lock. like below

`
/**
 * DMA context structure
 * Note: the dma_context shall be the first member
 *       of DMA client driver Data, got by dev->data
 *
 * magic			- magic code to identify the context
 * dma_channels		- dma channels
 * atomic			- driver atomic_t pointer
 *
 */
#define MAX_CHANNELS 64
struct dma_context {
	int32_t magic;
	int dma_channels;
	atomic_t *atomic;
       ATOMIC_DEFINE(channels_status_atomic, MAX_CHANNELS);
};

static inline int z_impl_dma_start(const struct device *dev, uint32_t channel)
{
	const struct dma_driver_api *api =
		(const struct dma_driver_api *)dev->api;
        struct dma_context *dma_ctx = (struct dma_context *)dev->data;

        if(atomic_test_and_set_bit(&dma_ctx->channels_status_atomic, channel)){
	        return api->start(dev, channel);
        } else {
                return 0;
        }
}

static inline int z_impl_dma_stop(const struct device *dev, uint32_t channel)
{
	const struct dma_driver_api *api =
		(const struct dma_driver_api *)dev->api;
       struct dma_context *dma_ctx = (struct dma_context *)dev->data;

       if(!atomic_test_and_clear_bit(&dma_ctx->channels_status_atomic, channel))
       {
               return 0;
       } else {
	        return api->stop(dev, channel);
      }
}

`

Reply:
Its an interesting idea, though in some cases today repeated start I think may mean different things.

Perhaps starting this out limiting this behavioral change to be repeated stop as being allowable is better and simpler. I believe that resolves the issue @tmleman ran into. It's also easier to test.